code/modules/paperwork/faxmachine.dm

```diff
@@ -1,235 +1,675 @@
+
+////////////////////////////////////////////////////////////////////////////////////////
+// Fax Machine Network Device
+////////////////////////////////////////////////////////////////////////////////////////
+/datum/extension/network_device/fax
+	has_commands = TRUE
+	long_range   = TRUE
+
+////////////////////////////////////////////////////////////////////////////////////////
+// Fax Machine Quick-Dial file
+////////////////////////////////////////////////////////////////////////////////////////
+/datum/computer_file/data/fax_quick_dial
+	filetype = "FQD"
+
+/datum/computer_file/data/fax_quick_dial/proc/save_quick_dial(var/list/quick_dial_list)
+	stored_data = json_encode(quick_dial_list)
+
+/datum/computer_file/data/fax_quick_dial/proc/load_quick_dial()
+	return json_decode(stored_data)
+
+////////////////////////////////////////////////////////////////////////////////////////
+// Fax Machine
+////////////////////////////////////////////////////////////////////////////////////////
+/obj/machinery/faxmachine
+	name                    = "fax machine"
+	icon                    = 'icons/obj/machines/fax_machine.dmi'
+	icon_state              = "fax"
+	anchored                = TRUE
+	density                 = TRUE
+	idle_power_usage        = 30
+	active_power_usage      = 200
+	construct_state         = /decl/machine_construction/default/panel_closed
+	maximum_component_parts = list(
+		/obj/item/stock_parts/disk_reader = 1,
+		/obj/item/stock_parts/card_reader = 1,
+		/obj/item/stock_parts/printer     = 1,
+		/obj/item/stock_parts             = 15,
+	)
+	stock_part_presets = list(
+		/decl/stock_part_preset/access_lock, //Empty access list
+	)
+	uncreated_component_parts = null
+	public_methods = list(
+		/decl/public_access/public_method/fax_receive_document,
+	)
+	var/obj/item/stock_parts/disk_reader/disk_reader  //Cached ref to the disk_reader component. Used for handling data disks.
+	var/obj/item/stock_parts/card_reader/card_reader  //Cached ref to the card_reader component. Used for scanning a user's id card.
+	var/obj/item/stock_parts/printer/printer          //Cached ref to the printer component. Used for printing things.
+	var/obj/item/scanner_item                         //Item to fax
+	var/list/quick_dial                              //List of name tag to network ids for other fax machines that the user added as quick dial options
+	var/list/fax_history                             //List of the last 10 devices that sent us faxes, and when
+
+	var/tmp/time_cooldown_end = 0
+	var/tmp/current_page      = "main"
+	var/tmp/dest_uri                   //The currently set destination URI for the target machine. Format is "[network_tag].[network_id]"
+
+/obj/machinery/faxmachine/Initialize(ml)
+	. = ..()
+	if(. != INITIALIZE_HINT_QDEL)
+		global.allfaxes += src
+		set_extension(src, /datum/extension/network_device/fax)
+		if(ml && printer)
+			printer.make_full()
+
+/obj/machinery/faxmachine/Destroy()
+	global.allfaxes -= src
 	. = ..()
```

Comment:
Those ref got a pretty well contained lifetime. When removing a part, it will call refresh_parts, which will clear the refs. It seems that qdeleting a part while inside a machine doesn't have any handling, which might be a bug, since then anything calculated from parts would be wrong. But that seems to be more of a wide-scale implementation oversight?

The main reason for those refs is that they're very much central to how the machine operates. And having to constantly fetch it from the components list when we could already determine if we had that thing or not on install/removal would get expensive after a whille. Especially when updating the UI via autoupdates and such. 

I'll clear them on faxmachine/Destroy too I guess.

Reply:
Machinery, on part install, registers for the part's destroyed event. When the part is destroyed, it removes references to the part. You have to do this too if you want to keep refs to parts like this.

When I say "handle GC" the issue is not what happens when the machine is destroyed--that's easy, the machine (presumably) is deleted cleanly and the refs it carries become irrelevant (though yes, it's certainly better to wipe them). It's what happens when the thing which is referenced is destroyed.

Reply:
> Machinery, on part install, registers for the part's destroyed event. When the part is destroyed, it removes references to the part. You have to do this too if you want to keep refs to parts like this.
> 
> When I say "handle GC" the issue is not what happens when the machine is destroyed--that's easy, the machine (presumably) is deleted cleanly and the refs it carries become irrelevant (though yes, it's certainly better to wipe them). It's what happens when the thing which is referenced is destroyed.

If they get deleted, and their event is called, its gonna call RefreshParts which will clear the ref to the part that was just removed anyways?

Reply:
I don't think they call RefreshParts, but I'm not sure.

Reply:
```
/obj/machinery/proc/component_destroyed(var/obj/item/component)
	events_repository.unregister(/decl/observ/destroyed, component, src)
	LAZYREMOVE(component_parts, component)
	LAZYREMOVE(processing_parts, component)
	power_components -= component
```

Apparently, they don't. But then that's a bug, and pretty messy. Component efficiency won't get recalculated, and uninstall code isn't run either. 
Honestly, the only way a component can be straight up qdeled inside a machine is either if the machine is deleted, or if someone manually VV and delete it.
Even explosion_act skip over components which are inside machines. I don't think its a very well supported or planned use case.

Reply:
@afterthought2